chore: Changed the process to expand yaml so that maxUnavailable, etc. can be used.

kahirokunn opened the pull request:
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #

<!--
**Is the PR title following semantic convention?**
Please refer to [Semantic types](https://github.com/open-policy-agent/gatekeeper/blob/master/.github/semantic.yml) to view accepted title convention to satisfy this status check.  
-->

<!--
**Are you making changes to Gatekeeper Helm chart?**
Please refer to [Contributing to Helm Chart](https://open-policy-agent.github.io/gatekeeper/website/docs/help#contributing-to-helm-chart) for modifying the Helm chart.
-->

<!--
**Are you making changes to Gatekeeper docs?**
Please see [Contributing to Docs](https://open-policy-agent.github.io/gatekeeper/website/docs/help#contributing-to-docs) 
-->

**Special notes for your reviewer**:

sozercan commented:
@kahirokunn thanks for the PR! Gatekeeper helm chart is autogenerated and these changes will be clobbered in the next release. Please refer to [Contributing to Helm Chart](https://open-policy-agent.github.io/gatekeeper/website/docs/help#contributing-to-helm-chart) for modifying the Helm chart.

kahirokunn commented:
@sozercan thx. done.

kahirokunn commented:
@sozercan Hello 👀

JaydipGabani commented:
@kahirokunn can you instead changing vairable `minAvailable`, introduce a new variable `maxUnavailable` and respective field in PDB? that way by not chaning original variable we are not breaking users(that are using `maxAvailable` variable).

kahirokunn commented:
/reopen

kahirokunn commented:
@JaydipGabani Introducing a new variable, maxUnavailable, instead of changing minAvailable is a thoughtful suggestion to avoid breaking existing users. However, minAvailable and maxUnavailable cannot be specified together as they are mutually exclusive. Adding maxUnavailable alone will not function correctly in the current setup.

kahirokunn commented:
@sozercan Hi 😢

JaydipGabani reviewed (approved):
lgtm

ritazh commented on manifest_staging/charts/gatekeeper/templates/gatekeeper-controller-manager-poddisruptionbudget.yaml:
```diff
@@ -12,7 +12,7 @@ metadata:
   namespace: '{{ .Release.Namespace }}'
 spec:
-  minAvailable: {{ .Values.pdb.controllerManager.minAvailable }}
+  minAvailable: {{ .Values.pdb.controllerManager | toYaml | nindent 2 }}
```
changes in this file should be generated by running “make manifests” for updates in [manifest_staging](https://github.com/open-policy-agent/gatekeeper/tree/master/manifest_staging) directory

kahirokunn replied:
okay. done.

ritazh reviewed (changes requested):
Thanks for the PR!

ritazh commented on cmd/build/helmify/replacements.go:
```diff
@@ -221,7 +221,7 @@ var replacements = map[string]string{
 	"HELMSUBST_MUTATING_WEBHOOK_MATCH_CONDITIONS": `{{ toYaml .Values.mutatingWebhookMatchConditions | nindent 4 }}`,
 
-	"HELMSUBST_PDB_CONTROLLER_MANAGER_MINAVAILABLE": `{{ .Values.pdb.controllerManager.minAvailable }}`,
+	"HELMSUBST_PDB_CONTROLLER_MANAGER": `{{ .Values.pdb.controllerManager | toYaml | nindent 2 }}`,
```
Unfortunately this does not ensure backward compat. Instead please add a new value for maxUnavailable and if/else block to ensure only one of them is set.


e.g. something like: 

{{- if and .Values.maxAvailable .Values.maxUnavailable }}
  {{ fail "Only one of 'maxAvailable' or 'maxUnavailable' can be set, but not both" }}
{{- else if not (or .Values.maxAvailable .Values.maxUnavailable) }}
  {{ fail "One of 'maxAvailable' or 'maxUnavailable' must be set" }}
{{- else }}
  # Continue with resource configuration
  {{ if .Values.maxAvailable }}
    maxAvailable: {{ .Values.maxAvailable }}
  {{ else if .Values.maxUnavailable }}
    maxUnavailable: {{ .Values.maxUnavailable }}
  {{ end }}
{{- end }}

kahirokunn replied:
okay. done.

kahirokunn commented on charts/gatekeeper/templates/gatekeeper-controller-manager-poddisruptionbudget.yaml:
```diff
@@ -11,8 +11,7 @@ metadata:
   namespace: '{{ .Release.Namespace }}'
-spec:
-  minAvailable: {{ .Values.pdb.controllerManager.minAvailable }}
+spec: {{ .Values.pdb.controllerManager | toYaml | nindent 2 }}
```
okay. done.

kahirokunn commented:
@ritazh thx for your review 🙏  how about now?

kahirokunn commented:
/refresh

kahirokunn commented:
@ritazh hi 👋

ritazh commented on cmd/build/helmify/kustomize-for-helm.yaml:
```diff
@@ -316,7 +316,7 @@ metadata:
   name: gatekeeper-controller-manager
   namespace: gatekeeper-system
 spec:
-  minAvailable: HELMSUBST_PDB_CONTROLLER_MANAGER_MINAVAILABLE
```
To ensure backward compatibility, please restore all existing use of PDB minAvailable. These are two different things and if there’s any conflict, kube-controller-manager will manage it.

kahirokunn replied:
Thank you for the suggestion. To validate the behavior, I tested a PDB that sets both fields:

```yaml
apiVersion: policy/v1
kind: PodDisruptionBudget
metadata:
  name: pdb-both-fields
  namespace: default
spec:
  minAvailable: 1
  maxUnavailable: 1
  selector:
    matchLabels:
      app: my-app
```

The API server rejects this with:

```
The PodDisruptionBudget "pdb-both-fields" is invalid: spec: Invalid value: {"MinAvailable":1,"Selector":{"matchLabels":{"app":"my-app"}},"MaxUnavailable":1,"UnhealthyPodEvictionPolicy":null}: minAvailable and maxUnavailable cannot be both set
```

Given this, kube-controller-manager never gets a chance to “resolve” a conflict—Kubernetes validation requires that only one of these fields be set. To preserve backward compatibility, the chart keeps minAvailable as the default and allows maxUnavailable as an opt-in. The rendered PDB includes exactly one of them (maxUnavailable if provided, otherwise minAvailable). If you prefer, we can also add a Helm validation that fails when both are set in values to surface the error earlier.

ritazh commented on config/pdb/pdb.yaml:
```diff
@@ -3,7 +3,6 @@ kind: PodDisruptionBudget
 metadata:
   name: controller-manager
 spec:
-  minAvailable: 1
```
please revert

ritazh replied:
I’m not sure if there is a way to preserve this for deploy/gatekeeper.yaml while enabling this new conditional to support minUnavailable and maxUnavailable in pdb. Do you @sozercan ?

ritazh commented on manifest_staging/charts/gatekeeper/templates/gatekeeper-controller-manager-poddisruptionbudget.yaml:
```diff
@@ -13,7 +13,17 @@ metadata:
   namespace: '{{ .Release.Namespace }}'
 spec:
-  minAvailable: {{ .Values.pdb.controllerManager.minAvailable }}
+  {{- if and .Values.maxAvailable .Values.maxUnavailable }}
```
These are incorrect.

maxUnavailble is part of the rolling update strategy part of the Deployment spec.

minAvailable is part of PDB.



e.g:

apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-web-app
spec:
  replicas: 10  # Ensure 10 running replicas
  strategy:
    type: RollingUpdate
    rollingUpdate:
      maxUnavailable: 2  # Allow at most 2 pods to be unavailable at any time during updates
  template:
    metadata:
      labels:
        app: my-web-app
    spec:
      containers:
        - name: my-web-app
          image: my-web-app:v2

Separately, PDB:

apiVersion: policy/v1
kind: PodDisruptionBudget
metadata:
  name: my-web-app-pdb
spec:
  minAvailable: 7  # Ensure at least 7 pods are always running
  selector:
    matchLabels:
      app: my-web-app

ritazh replied:
Suggestion: revert PDB minAvailable changes, add support for maxUnavailable rolling update strategy, update helm chart values yaml,  update helm chart doc, generate chart changes.

kahirokunn replied:
Thank you for the swift review and the detailed feedback — I really appreciate it.

Just to clarify: the field I’m discussing in this PR is **`spec.maxUnavailable` on a PodDisruptionBudget (PDB)**, not the `strategy.rollingUpdate.maxUnavailable` field on a Deployment. In a PDB, `maxUnavailable` and `minAvailable` are mutually exclusive ways to express the disruption policy (see the “Specifying a PodDisruptionBudget” section of the Kubernetes documentation and `kubectl explain poddisruptionbudget.spec`).
https://kubernetes.io/docs/tasks/run-application/configure-pdb/#specifying-a-poddisruptionbudget

Could you let me know if this explanation addresses your concern, or if there’s anything else I should adjust?

Thanks again for your time and insight! 🙏

ritazh replied:
You are right. Thanks!

kahirokunn commented:
@ritazh I would appreciate it if you could take a look at this when you have time 🙏

kahirokunn commented on manifest_staging/charts/gatekeeper/templates/gatekeeper-controller-manager-poddisruptionbudget.yaml:
```diff
@@ -13,7 +13,17 @@ metadata:
+  {{- if and .Values.maxAvailable .Values.maxUnavailable }}
+    {{ fail "Only one of 'maxAvailable' or 'maxUnavailable' can be set, but not both" }}
+    {{- else if not (or .Values.maxAvailable .Values.maxUnavailable) }}
+    {{ fail "One of 'maxAvailable' or 'maxUnavailable' must be set" }}
```
Thx. I fixed.

kahirokunn commented:
Hi @ritazh, thank you for your earlier feedback. I’ve updated the PDB templates to support maxUnavailable while preserving the existing minAvailable behavior, and updated values/README accordingly. When you have a moment, could you please review again? Thank you!
